### PR TITLE
feat: built-in function development guide

### LINKED ISSUE #781: OpenMLDB Build-In Function Develop Guide
**Is your feature request related to a problem? Please describe.**

## Backgroud

OpenMLDB contains hundreds of build-in functions helping data scientist to exetract features and analyze data. For example, now we have aggregation functions like `SUM`, `AVG`, `MAX`, `MIN`, `COUNT`, etc,  to aggregate data over a table or over a specific window. In addition to that, we also have scalar functions like `MINUTE`, `HOUR`, `SECOND`, `SIN`, `COS`, `LOG`, etc, to extract features based on one-row data. 

The article is a hands-on guide to developing built-in functions in OpenMLDB. 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by jingchen2222**
* **Please check if the PR fulfills these requirements**

- The commit message follows our guidelines
- Tests for the changes have been added (for bug fixes / features)
- Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
docs update

* **What is the current behavior?** (You can also link to an open issue here)
[* ](https://github.com/4paradigm/OpenMLDB/issues/781)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

**Comment by jingchen2222** [comment]
[UDF development docs](https://trello.com/c/h29h0rH0/10-udf-development-docs)

File: docs/cn/built_in_function_develop_guide.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+  - 数学函数
+  - 逻辑函数
+  - 日期函数
+  - 字符串函数
+
+本文将介绍单行函数的开发，暂不深入更为复杂的聚合函数开发细节。
+
+## 开发和注册内置函数
+
+`ExternalFuncRegistryHelper` is OpenMLDB的内置函数开发最重要的辅助类。它提供一套API来配置函数，并将配置好的函数注册到相应的函数库中。完成函数注册后，可以在SQL查询语句中调用函数。本章节主要介绍内置函数的注册和配置的基本步骤和并提供若干示范。
```
Comment (zhanghaohit) [inline docs/cn/built_in_function_develop_guide.md]: is -> 是

Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md]: add link to `ExternalFuncRegistryHelper`

File: docs/cn/built_in_function_develop_guide.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+### ExternalFuncRegistryHelper API
+
+OpenMLDB内置的函数都应该注册到默认函数库`DefaultUdfLibrary`(src/udf/default_udf_library.cc)中。`DefaultUdfLibrary`提供`RegisterExternal` 接口来创建一个函数注册助手。它负责配置函数的参数，配置返回值了行，并注册函数到默认函数库中。
```
Comment (zhanghaohit) [inline docs/cn/built_in_function_develop_guide.md]: 配置返回值了行?

File: docs/cn/built_in_function_develop_guide.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+-  `return_by_arg()`  : 配置结果是否通过参数返回
+  - 当 **return_by_arg(false)**时 , 结果直接通过`return`返回. OpenMLDB 默认配置  `return_by_arg(false) ` .
+  - 当 **return_by_arg(true)** 时,结果通过参数返回。
+    - 若返回类型是***non-nullable***, 函数结果将通过最后一个参数返回。
+    - 若返回类型是**nullable**, 函数结果值将通过倒数第二个参数返回，而 ***null flag*** 将通过最后一个参数返回。如果***null flag***为***true***, 那么函数结果为***null***，否则函数结果从倒数第二个参数读取。
+- `doc()`: 配置函数文档
+
+### Register built-in function `return_by_arg(false)`
+
+一般来说，当函数的结果是一个数值或者bool值时，我们配置`return_by_arg` as `false` 。那意味着，函数结果将直接return返回。
```
Comment (zhanghaohit) [inline docs/cn/built_in_function_develop_guide.md]: as -> 为

File: docs/cn/built_in_function_develop_guide.md (hunk 2 above)
Comment (zhanghaohit) [reply docs/cn/built_in_function_develop_guide.md]: src/udf/default_udf_library.cc这个要加上hybridse么？

hybridse/src/udf/default_udf_library.cc

File: docs/cn/built_in_function_develop_guide.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  .returns<return_type>
+  .doc(documentation)
+```
+
+- `RegisterExternal(function_name)`: 可以创建一个`ExternalFuncRegistryHelper`对象，初始化函数的注册名。SQL使用这个名字来调用函数。
+- `built_in_fn_pointer`: 被注册的C++函数指针
+- `args<arg_type,...>`: 配置函数参数类型
+-  `returns<return_type>`: 配置函数返回值类型
+-  `return_by_arg()`  : 配置结果是否通过参数返回
+  - 当 **return_by_arg(false)**时 , 结果直接通过`return`返回. OpenMLDB 默认配置  `return_by_arg(false) ` .
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md]: Indentation and bold do not work

File: docs/cn/built_in_function_develop_guide.md [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    - 若返回类型是**nullable**, 函数结果值将通过倒数第二个参数返回，而 ***null flag*** 将通过最后一个参数返回。如果***null flag***为***true***, 那么函数结果为***null***，否则函数结果从倒数第二个参数读取。
+- `doc()`: 配置函数文档
+
+### Register built-in function `return_by_arg(false)`
+
+一般来说，当函数的结果是一个数值或者bool值时，我们配置`return_by_arg` as `false` 。那意味着，函数结果将直接return返回。
+
+#### step 1: 实现待注册的内置函数
+
+一般地，开发者可以在(`src/udf/udf.h` 和  `src/udf/udf.cc`)中实现内置函数。
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md]: declar in `udf.h` and implement in `udf.cc`

File: docs/cn/built_in_function_develop_guide.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+#### step 1: 实现待注册的内置函数
+
+**示例：**
+
+我们在`udf.cc`中实现两个`timestamp_to_date()`函数。timestamp_to_date接受一个`timestamp`参数，并将`timestamp`转成`date`。由于OpenMLDB的`date`类型是一个结构体类型，设计函数是，不直接返回结果，而是将结果存放在参数中返回。与此同时，考虑到日期转换可能有异常或失败，转换失败是，应返回null。所以，我们额外增加一个null flag作为最后一个参数。
+
+```c++
+namespace v1 {
+  void timestamp_to_date(codec::Timestamp *timestamp,
+       									 codec::Date *output /*result output*/, bool *is_null /*null flag*/) {
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md]: The parameter of function displayed not very well

File: docs/cn/built_in_function_develop_guide.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+## 函数文档配置
+
+`ExternalFuncRegistryHelper` 提供了`doc(doc_string)`API来配置文档。函数文档用来描述函数的功能和用法。它的主要受众是普通用户。所以，文档需要清晰易懂。
+
+函数文档一般需要包含一下几类信息：
+
+- **@brief** ：简要描述函数的功能
+- **@param**：函数参数描述
+- **Examples**： 函数的使用示例。一般需要提供SQL的查询示例。SQL语句需要放置在 a `@code/@endcode` 代码块中。
+- **@since** ：指明引入该函数的OpenMLDB版本
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md]: Where to get the next release version

File: docs/cn/built_in_function_develop_guide.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+}
+// date(timestamp) normal check
+TEST_F(UdfIRBuilderTest, timestamp_to_date_test_0) {
+    CheckUdf<Nullable<Date>, Nullable<Timestamp>>(
+        "date", codec::Date(2020, 05, 20), codec::Timestamp(1589958000000L));
+}
+// date(timestamp) null check
+TEST_F(UdfIRBuilderTest, timestamp_to_date_test_null_0) {
+    CheckUdf<Nullable<Date>, Nullable<Timestamp>>("date", nullptr, nullptr);
+}
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md:730]: How to test end2end with sql cmd, link to the compile doc?

File: docs/cn/built_in_function_develop_guide.md [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+## 函数文档配置
+
+`ExternalFuncRegistryHelper` 提供了`doc(doc_string)`API来配置文档。函数文档用来描述函数的功能和用法。它的主要受众是普通用户。所以，文档需要清晰易懂。
+
+函数文档一般需要包含一下几类信息：
+
+- **@brief** ：简要描述函数的功能
+- **@param**：函数参数描述
+- **Examples**： 函数的使用示例。一般需要提供SQL的查询示例。SQL语句需要放置在 a `@code/@endcode` 代码块中。
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md]: delete `a`

File: docs/cn/built_in_function_develop_guide.md [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+函数名和函数参数的配置和普通函数配置一样。但需要额外注意返回值类型的配置：
+
+- 因为函数结果存放在参数中返回，所以配置`return_by_arg(true)`
+- 因为函数结果可能为null,所以配置`.returns<Nullable<Date>>`
+
+```c++
+RegisterExternal("date")
+        .args<codec::Timestamp>(reinterpret_cast<void*>(
+            static_cast<void (*)(Timestamp*, Date*, bool*)>(
```
Comment (zhanghaohit) [inline docs/cn/built_in_function_develop_guide.md]: why we need both `reinterpret_cast` and `static_cast` here?

File: docs/cn/built_in_function_develop_guide.md [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+RegisterExternal(function_name)
+  .args<arg_type, ...>(built_in_fn_pointer)
+  .return_by_arg(bool_value)
+  .returns<return_type>
+  .doc(documentation)
+```
+
+- `RegisterExternal(function_name)`: 可以创建一个`ExternalFuncRegistryHelper`对象，初始化函数的注册名。SQL使用这个名字来调用函数。
+- `built_in_fn_pointer`: 被注册的C++函数指针
+- `args<arg_type,...>`: 配置函数参数类型
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md]: It is better to describe the type of function parameter. ex: the difference between `hybridse::codec::StringRef`  and  `std::string` , and how to select

File: docs/cn/built_in_function_develop_guide.md [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+```c++
+RegisterExternal(function_name)
+  .args<arg_type, ...>(built_in_fn_pointer)
+  .return_by_arg(bool_value)
+  .returns<return_type>
+  .doc(documentation)
+```
+
+- `RegisterExternal(function_name)`: 可以创建一个`ExternalFuncRegistryHelper`对象，初始化函数的注册名。SQL使用这个名字来调用函数。
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md]: the format of function_name and parameter_name format, lower or lower camel case.

File: docs/cn/built_in_function_develop_guide.md [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+- `built_in_fn_pointer`: 被注册的C++函数指针
+- `args<arg_type,...>`: 配置函数参数类型
+-  `returns<return_type>`: 配置函数返回值类型
+-  `return_by_arg()`  : 配置结果是否通过参数返回
+  - 当 **return_by_arg(false)**时 , 结果直接通过`return`返回. OpenMLDB 默认配置  `return_by_arg(false) ` .
+  - 当 **return_by_arg(true)** 时,结果通过参数返回。
+    - 若返回类型是***non-nullable***, 函数结果将通过最后一个参数返回。
+    - 若返回类型是**nullable**, 函数结果值将通过倒数第二个参数返回，而 ***null flag*** 将通过最后一个参数返回。如果***null flag***为***true***, 那么函数结果为***null***，否则函数结果从倒数第二个参数读取。
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md]: There are different functions where `RegisterExternal`  should be registered, ex `IniMathUdf`, `InitStringUdf`.  It need to select by function type

File: docs/cn/built_in_function_develop_guide.md [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+    CheckUdf<int32_t, Timestamp>("month", 5, time);
+}
+// date(timestamp) normal check
+TEST_F(UdfIRBuilderTest, timestamp_to_date_test_0) {
+    CheckUdf<Nullable<Date>, Nullable<Timestamp>>(
+        "date", codec::Date(2020, 05, 20), codec::Timestamp(1589958000000L));
+}
+// date(timestamp) null check
+TEST_F(UdfIRBuilderTest, timestamp_to_date_test_null_0) {
+    CheckUdf<Nullable<Date>, Nullable<Timestamp>>("date", nullptr, nullptr);
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md:729]: How to build and run the ut.

File: docs/cn/built_in_function_develop_guide.md (hunk 8 above)
Comment (jingchen2222) [reply docs/cn/built_in_function_develop_guide.md:730]: I will add a section for:
# Integration Test Case

File: docs/cn/built_in_function_develop_guide.md (hunk 11 above)
Comment (jingchen2222) [reply docs/cn/built_in_function_develop_guide.md]: I will add a section like:
### How to Implement a built-in function
here,
I will describe the type mapping from OpenMLDB type to c++ type.

File: docs/cn/built_in_function_develop_guide.md (hunk 12 above)
Comment (jingchen2222) [reply docs/cn/built_in_function_develop_guide.md]: Sure. I will statement that function and parameter names are case insensitivity

File: docs/cn/built_in_function_develop_guide.md (hunk 13 above)
Comment (jingchen2222) [reply docs/cn/built_in_function_develop_guide.md]: Yes. I will statement that.

**Comment by jingchen2222** [comment]
- [x] Add Section for `RegisterAlias`

File: docs/cn/built_in_function_develop_guide.md (hunk 2 above)
Comment (jingchen2222) [reply docs/cn/built_in_function_develop_guide.md]: done

File: docs/cn/built_in_function_develop_guide.md [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+#### step 2: 配置函数，并注册到默认函数库中
+
+使用`ExternalFuncRegistryHelper`工具类配置参数和注册函数到默认库中。
+
+**示例：**
+
+我们将前一节中实现的两个c++函数注册到默认库中。
+
+```c++
+RegisterExternal("month")
```
Comment (zhanghaohit) [inline docs/cn/built_in_function_develop_guide.md]: udf.cc not use `RegisterExternal` to register function? I saw it is using `DefaultUdfLibrary::get()->AddExternalFunction`? 

Maybe we would better to prepare an end2end example just like the docs shows.

File: docs/cn/built_in_function_develop_guide.md (hunk 11 above)
Comment (jingchen2222) [reply docs/cn/built_in_function_develop_guide.md]: done

File: docs/cn/built_in_function_develop_guide.md [hunk 16]
```diff
@@ -0,0 +1,590 @@
+# OpenMLDB SQL 内置函数开发指南
+
+## 背景
```
Comment (lumianph) [inline docs/cn/built_in_function_develop_guide.md]: add section numbers to all section titles. The font size cannot easily distinguish the sections.

such as 
1. 背景
2. 开发和注册内置函数
2.1. 开发 C++ 内置函数 
2.2. 注册C++函数到默认函数库中

File: docs/cn/built_in_function_develop_guide.md [hunk 17]
```diff
@@ -0,0 +1,590 @@
...
+  } // namespace v1
+} // namespace udf
+```
+
+**step 2: 配置函数，并注册到默认函数库中**
+
+使用`ExternalFuncRegistryHelper`工具类配置参数和并注册函数到默认库中:
+
+```c++
+RegisterExternal("month")
```
Comment (lumianph) [inline docs/cn/built_in_function_develop_guide.md]: where to add this piece of code (the file path)?

File: docs/cn/built_in_function_develop_guide.md [hunk 18]
```diff
@@ -0,0 +1,590 @@
...
+```SQL
+select MONTH(TIMESTAMP(1590115420000)) as m1, month(timestamp(1590115420000)) as m2;
+ ---- ---- 
+  m1   m2  
+ ---- ---- 
+  5    5  
+ ---- ---- 
+```
+
+### 场景2: 函数结果通过参数返回，结果不为null
```
Comment (lumianph) [inline docs/cn/built_in_function_develop_guide.md]: refer to the comment of example 1

File: docs/cn/built_in_function_develop_guide.md [hunk 19]
```diff
@@ -0,0 +1,590 @@
...
+  - 当 **return_by_arg(false)**时 , 结果直接通过`return`返回. OpenMLDB 默认配置  `return_by_arg(false) ` 
+  - 当 **return_by_arg(true)** 时,结果通过参数返回
+    - 若返回类型是***non-nullable***, 函数结果将通过最后一个参数返回
+    - 若返回类型是**nullable**, 函数结果值将通过倒数第二个参数返回，而 ***null flag*** 将通过最后一个参数返回。如果***null flag***为***true***, 那么函数结果为***null***，否则函数结果从倒数第二个参数读取
+- `doc()`: 配置函数文档。文档配置请遵循[函数文档配置](#函数文档配置)的规范
+
+### 场景1: 函数结果直接返回 `return_by_arg(false)`
+
+一般来说，当SQL函数返回值为布尔或数值类型当函数的结果是一个数值(***bigint, int, smallint, float, double***) 或者***bool***值时，可以将结果直接return返回。可以将函数设计为如下的样子：
+
```
Comment (lumianph) [inline docs/cn/built_in_function_develop_guide.md:193]: you may consider removing the template code, go to the example directly

**Review by lumianph — CHANGES_REQUESTED** [review]

File: docs/cn/built_in_function_develop_guide.md [hunk 20]
```diff
@@ -0,0 +1,590 @@
...
+# hybridse/src/udf/udf.h
+namespace udf{
+  namespace v1 {
+    int32_t month(int64_t ts);
+    int32_t month(codec::Timestamp *ts);
+  } // namespace v1
+} // namespace udf
+```
+
+在[hybridse/src/udf/udf.cc](https://github.com/4paradigm/OpenMLDB/blob/main/hybridse/src/udf/udf.cc)中实现`month()`函数:
```
Comment (zhanghaohit) [inline docs/cn/built_in_function_develop_guide.md]: 可以直接`(hybridse/src/udf/udf.cc)`，不同branch上会对应不同的代码

File: docs/cn/built_in_function_develop_guide.md [hunk 21]
```diff
@@ -0,0 +1,647 @@
...
+
+注册成功后，用户在SQL语句中使用这个名字来调用函数。请注意，SQL中的函数大小写不敏感。
+
+```SQL
+SELECT register_func(col1) FROM t1;
+```
+
+#### 2.2.2 配置函数参数和返回值
+
+`ExternalFuncRegistryHelper` is OpenMLDB的内置函数开发最重要的辅助类。它提供一套API来配置函数，并将配置好的函数注册到相应的函数库中。
```
Comment (lumianph) [inline docs/cn/built_in_function_develop_guide.md]: is => 是

File: docs/cn/built_in_function_develop_guide.md [hunk 22]
```diff
@@ -0,0 +1,647 @@
...
+        void func_output_str(int32_t, codec::StringRef*); 
+        ```
+
+    - 当SQL函数的返回值可能为空(**Nullable**)时，额外要有一个`bool*`类型的参数来存放结果为空与否
+
+      - ```c++
+        // SQL: Nullable<DATE> FUNC_NULLABLE_DATE(BIGINT)
+        void func_output_nullable_date(int64_t, codec::Date*, bool*); 
+        ```
+
```
Comment (lumianph) [inline docs/cn/built_in_function_develop_guide.md:86]: 你可以在这里结束的时候加类似一句话来引出后面为什么要分三种情况讨论：
注意， SQL 函数返回值将对内置函数的实现和注册方式产生较大的影响，我们将在后续分别讨论，详情参见 “3. 几种SQL函数的开发流程详细介绍” 。。

File: docs/cn/built_in_function_develop_guide.md [hunk 23]
```diff
@@ -0,0 +1,647 @@
...
+```bash
+cd ./hybridse
+mkdir -p build
+cd build
+cmake .. -DCMAKE_BUILD_TYPE=Release -DTESTING_ENABLE=ON
+make udf_ir_builder_test -j4
+SQL_CASE_BASE_DIR=${OPENMLDB_DIR} ./src/codegen/udf_ir_builder_test
+```
+
+## 3. 几种SQL函数的开发流程详细介绍
```
Comment (lumianph) [inline docs/cn/built_in_function_develop_guide.md]: SQL 函数开发模板 （？）

File: docs/cn/built_in_function_develop_guide.md [hunk 24]
```diff
@@ -0,0 +1,647 @@
...
+RegisterExternal("my_func")
+        .args<Arg1, Arg2, ...>(static_cast<R (*)(Arg1, Arg2, ...)>(v1::func))
+  			.return_by_arg(true)
+  			.returns<Nullable<Ret>>()
+        .doc(R"(
+            documenting my_func
+        )");
+```
+
+## 4. 几种SQL函数开发的实例
```
Comment (lumianph) [inline docs/cn/built_in_function_develop_guide.md]: SQL 函数开发实例（？）

File: docs/cn/built_in_function_develop_guide.md (hunk 20 above)
Comment (jingchen2222) [reply docs/cn/built_in_function_develop_guide.md]: we might use this document in another docs project repository. So had better keep the repo host

**Review by lumianph — COMMENTED** [review]
you may make the example title to correspond to the template categories
e.g. 
4.1 SQL 函数返回值为布尔或数值类型 - INT Month(TIMESTAMP)

File: docs/cn/built_in_function_develop_guide.md [hunk 25]
```diff
@@ -0,0 +1,709 @@
...
+
+### 2.2. 注册和配置接口
+
+#### 2.2.1 注册位置
+
+OpenMLDB的 `DefaultUdfLibrary` 负责存放和管理内置的全局SQL函数。默认函数库`DefaultUdfLibrary` 声明在文件[hybridse/src/udf/default_udf_library.h](https://github.com/4paradigm/OpenMLDB/blob/main/hybridse/src/udf/default_udf_library.h)文件中，实现在[hybridse/src/udf/default_udf_library.cc](https://github.com/4paradigm/OpenMLDB/blob/main/hybridse/src/udf/default_udf_library.cc)文件中。
+
+开发者先在[hybridse/src/udf/default_udf_library.cc](https://github.com/4paradigm/OpenMLDB/blob/main/hybridse/src/udf/default_udf_library.cc)中将C++函数注册到默认函数库 `DefaultUdfLibrary` 中，用户才能通过SQL语句访问这个函数。为了更好的管理函数注册，`DefaultUdfLibrary`提供几类`DefaultUdfLibrary::InitXXXXUdf()` 接口来管理不同功能的函数配置注册:
+
+- 数学函数：在`DefaultUdfLibrary::IniMathUdf()`中注册
```
Comment (dl239) [inline docs/cn/built_in_function_develop_guide.md]: I have renamed this function to `InitMathUdf` in https://github.com/4paradigm/OpenMLDB/pull/791

**Comment by lumianph** [comment]
the Chinese version looks good to me, you may revise the English version accordingly.

**Review by lumianph — APPROVED** [review]
let's keep the English version for now (overall the doc is in good shape). I am planning to deliver a tutorial about English doc writing, probably use this doc as an example :-)

**Comment by lumianph** [comment]
@jingchen2222 you can merge this PR now, good effort 👍

File: docs/cn/built_in_function_develop_guide.md (hunk 25 above)
Comment (jingchen2222) [reply docs/cn/built_in_function_develop_guide.md]: Resolved

**Review by dl239 — APPROVED** [review]

File: docs/en/built_in_function_develop_guide.md [hunk 26]
```diff
@@ -0,0 +1,736 @@
...
+- Function names would be clear and readable. Use names that describe the purpose or intent of the function.
+
+#### 2.1.3 C++ and SQL Data Type
+
+C++ built-in function can use limited data types, including BOOL, Numeric, String, Timestamp and Date.The correspondence between the SQL data type and the C++ data type is shown as followed:
+
+- | SQL Type  | C++ Type           |
+  | :-------- | :----------------- |
+  | BOOL      | `bool`             |
+  | SMALLINT  | int16_t            |
```
Comment (dl239) [inline docs/en/built_in_function_develop_guide.md:51]: add `

File: docs/en/built_in_function_develop_guide.md [hunk 27]
```diff
@@ -0,0 +1,736 @@
...
+  | DOUBLE    | `double`           |
+  | STRING    | `codec::StringRef` |
+  | TIMESTAMP | `codec::Timestamp` |
+  | DATE      | `codec::Date`      |
+
+#### 2.1.4 Parameters and Result
+
+- SQL function parameters and C++ function parameters have the same position order.
+
+- C++ function Parameter types should be match with SQL type. Check [2.1.3 C++ and SQL Data Type](#2.1.3-C++ and SQL Data Type) for details.
```
Comment (dl239) [inline docs/en/built_in_function_develop_guide.md:64]: cannot jump

File: docs/en/built_in_function_develop_guide.md [hunk 28]
```diff
@@ -0,0 +1,736 @@
...
+      ```
+
+  - If SQL function return type is ***Nullable***, we need one more `bool*`parameter to return a `is_null` flag.
+
+    - ```c++
+      // SQL: Nullable<DATE> FUNC_NULLABLE_DATE(BIGINT)
+      void func_output_nullable_date(int64_t, codec::Date*, bool*); 
+      ```
+
+  - Notice that return types have greate impart on built-in function developing behaviours. We will discuss the details in later section [3. Built-in Function Development Template](#3.-Built-in Function Development Template).
```
Comment (dl239) [inline docs/en/built_in_function_develop_guide.md:89]: `impart` -> `impact`

File: docs/en/built_in_function_develop_guide.md [hunk 29]
```diff
@@ -0,0 +1,736 @@
...
+
+#### 2.1.5 Memory Management
+
+- Operator `new` operator or method `malloc` are forbidden in C++ built-in function implementation.
```
Comment (dl239) [inline docs/en/built_in_function_develop_guide.md:93]: duplicated operator

**Comment by jingchen2222** [comment]
[docs reorganize, review and udpate](https://trello.com/c/EW0F7yOb/1-docs-reorganize-review-and-udpate)

**Pull request merged**